lib/ansible/modules/cloud/vmware/vmware_export_ovf.py

```diff
@@ -0,0 +1,328 @@
+description: >
+   This module can be used to export a VMware virtual machine to OVF template from vCenter server or ESXi host.
+version_added: '2.8'
+author:
+- Diane Wang (@Tomorrow9) <dianew@vmware.com>
+requirements:
+- python >= 2.6
+- PyVmomi
+notes: []
+options:
+  name:
+    description:
+    - Name of the virtual machine to export.
+    - This is a required parameter, if parameter C(uuid) is not supplied.
+  uuid:
+    description:
+    - Uuid of the virtual machine to export.
+    - This is a required parameter, if parameter C(name) is not supplied.
+  datacenter:
+    default: ha-datacenter
+    description:
+    - Datacenter name of the virtual machine to export.
+    - This parameter is case sensitive.
+  folder:
+    description:
+    - Destination folder, absolute path to find the specified guest.
+    - The folder should include the datacenter. ESX's datacenter is ha-datacenter.
+    - This parameter is case sensitive.
+    - 'If multiple machines are found with same name, this parameter is used to identify
+       uniqueness of the virtual machine. version_added 2.5'
+    - 'Examples:'
+    - '   folder: /ha-datacenter/vm'
+    - '   folder: ha-datacenter/vm'
+    - '   folder: /datacenter1/vm'
+    - '   folder: datacenter1/vm'
+    - '   folder: /datacenter1/vm/folder1'
+    - '   folder: datacenter1/vm/folder1'
+    - '   folder: /folder1/datacenter1/vm'
+    - '   folder: folder1/datacenter1/vm'
+    - '   folder: /folder1/datacenter1/vm/folder2'
+  export_dir:
+    description:
+    - Absolute path to place the exported files on the server running this task, must have write permission.
+    - If folder not exist will create it, also create a folder under this path named with VM name.
+    required: yes
+  export_with_images:
+    default: false
+    description:
+    - Export an ISO image of the media mounted on the CD/DVD Drive within the virtual machine.
+    type: bool
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+- vmware_export_ovf:
+    validate_certs: false
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    name: '{{ vm_name }}'
+    export_with_images: true
+    export_dir: /path/to/ovf_template/
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+instance:
+    description: list of the exported files, if exported from vCenter server, device file is not named with vm name
+    returned: always
+    type: dict
+    sample: None
+'''
+
+import os
+import hashlib
+from time import sleep
+from threading import Thread
+from ansible.module_utils.urls import open_url
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.vmware import (connect_to_api, vmware_argument_spec, PyVmomi)
+try:
+    from pyVmomi import vim
+    from pyVim import connect
+except ImportError:
+    pass
+
+
+class LeaseProgressUpdater(Thread):
```

Comment:
"HttpNfcLease" object returned from "ExportVApp", "ExportVm" and "ImportVApp". "vmware_deploy_ovf" module already use "ImportVApp" function. 
And this module "vmware_export_ovf" using "ExportVm" function. One is upload, the other is download. If so the common part of code will not be called frequently, except when there is new requirement of "ExportVApp". So I'm putting the code in this module.

Reply:
I see the code in vmware_deploy_ovf and it could use this class as well... well the next time a lease needs to be managed, invoke the "Rule of Three" and refactor this into module_utils?

Reply:
Yeah, I think if there is new requirement of "ExportVApp". Then there is no chance to use it.